StakingRewards/contracts/StakingRewards.sol

```diff
@@ -0,0 +1,306 @@
+  /**
+   * Emitted when owner recovered tokens from this contract
+   */
+  event Recovered(address token, uint256 amount);
+
+  /* -------------------------------------------------------------------------- */
+  /*                                  Modifier                                  */
+  /* -------------------------------------------------------------------------- */
+
+  modifier updateReward(address account) {
+    rewardPerTokenStored = _rewardPerToken();
+    lastUpdateTime = _lastTimeRewardApplicable();
+    if (account != address(0)) {
+      rewards[account] = _earned(account);
+      userRewardPerTokenPaid[account] = rewardPerTokenStored;
+    }
+    _;
+  }
+
+  /* -------------------------------------------------------------------------- */
+  /*                             External Functions                             */
+  /* -------------------------------------------------------------------------- */
+
+  /**
+   * @dev Constructor
+   * @param rewardsToken_ Address to rewards ERC20 token
+   * @param stakingToken_ Address to staking ERC20 token
+   * @param duration Reward duration in seconds
+   */
+  constructor(address rewardsToken_, address stakingToken_, uint256 duration) {
+    rewardsToken = IERC20(rewardsToken_);
+    stakingToken = IERC20(stakingToken_);
+    rewardsDuration = duration;
+  }
+
+  /**
+   * @notice Set rewards duration, only available to set after finish
+   * of previous rewards period.
+   * @dev Callable by owner
+   * @param duration New rewards duration
+   */
+  function setRewardsDuration(uint256 duration) external onlyOwner {
+    if (block.timestamp < periodFinish) {
+      revert WaitToFinish();
+    }
+    rewardsDuration = duration;
+    emit RewardsDurationUpdated(duration);
+  }
+
+  /**
+   * @notice Stake staking token, callable only when unpaused.
+   * Every staked balance per user will be accumulated and it transfers tokens
+   * from user to this contract.
+   * @dev Callable only when unpaused
+   * @param amount Staking amount
+   */
+  function stake(
+    uint256 amount
+  ) public nonReentrant whenNotPaused updateReward(msg.sender) {
+    if (amount <= 0) {
+      revert ZeroAmount();
+    }
+    totalSupply = totalSupply + amount;
+    balances[msg.sender] += amount;
+    stakingToken.safeTransferFrom(msg.sender, address(this), amount);
+    emit Staked(msg.sender, amount);
+  }
+
+  /**
+   * @notice Unstake staking token, callable only when unpaused.
+   * Staked balance per user will be reduced and it transfers tokens back from
+   * this contract to the user.
+   * @dev Callable only when unpaused
+   * @param amount Unstaking amount
+   */
+  function unstake(
+    uint256 amount
+  ) public nonReentrant whenNotPaused updateReward(msg.sender) {
+    if (amount <= 0) {
+      revert ZeroAmount();
+    }
+    if (balances[msg.sender] < amount) {
+      revert NotEnoughBalance();
+    }
+    totalSupply = totalSupply - amount;
```

Comment:
you can change this line like this.
`totalSupply -= amount;`

Reply:
`totalSupply = totalSupply - amount;` is more gasless, @solcmoon